Description From Larry Finger 2008-12-20 10:45 [reply] 
Latest working kernel version: v2.6.27
Earliest failing kernel version: v2.6.28-rc1
Distribution: openSUSE 11.1
Hardware Environment: HP dv2815nv laptop
Software Environment: x86_64
Problem Description: When using LM87 firmware, the p54usb device stalls for
certain message lengths because the firmware needs an explicit termination
packet.

The fix is the following patch from Christian Lamparter <chunkeey@web.de>:

--- a/drivers/net/wireless/p54/p54usb.c
+++ b/drivers/net/wireless/p54/p54usb.c
@@ -251,6 +251,7 @@ static void p54u_tx_lm87(struct ieee8021
 		usb_sndbulkpipe(priv->udev, P54U_PIPE_DATA), hdr,
 		len + sizeof(*hdr), free_on_tx ? p54u_tx_free_cb : p54u_tx_cb,
 		dev);
+	data_urb->transfer_flags |= URB_ZERO_PACKET;
 
 	usb_submit_urb(data_urb, GFP_ATOMIC);
 }
